e2e/src/test/java/io/kubernetes/client/e2e/extended/leaderelection/LeaderElectorTest.java

```diff
@@ -0,0 +1,334 @@
+You may obtain a copy of the License at
+http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package io.kubernetes.client.e2e.extended.leaderelection;
+
+import io.kubernetes.client.extended.leaderelection.LeaderElectionConfig;
+import io.kubernetes.client.extended.leaderelection.LeaderElector;
+import io.kubernetes.client.extended.leaderelection.Lock;
+import io.kubernetes.client.extended.leaderelection.resourcelock.ConfigMapLock;
+import io.kubernetes.client.extended.leaderelection.resourcelock.EndpointsLock;
+import io.kubernetes.client.extended.leaderelection.resourcelock.LeaseLock;
+import io.kubernetes.client.openapi.ApiClient;
+import io.kubernetes.client.openapi.ApiException;
+import io.kubernetes.client.openapi.apis.CoordinationV1Api;
+import io.kubernetes.client.openapi.apis.CoreV1Api;
+import io.kubernetes.client.util.ClientBuilder;
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.BrokenBarrierException;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.CyclicBarrier;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import org.joda.time.format.*;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@RunWith(Parameterized.class)
+public class LeaderElectorTest {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(LeaderElectorTest.class);
+
+  private static final String LOCK_RESOURCE_NAME = "leader-election-it";
+  private static final String NAMESPACE = "default";
+
+  private enum LockType {
+    ConfigMap,
+    Endpoints,
+    Lease
+  }
+
+  @Parameterized.Parameters(name = "{0}")
+  public static Collection<Object[]> constructorFeeder() {
+    final List<Object[]> args = new ArrayList<>();
+
+    args.add(new Object[] {LockType.ConfigMap});
+    args.add(new Object[] {LockType.Endpoints});
+    args.add(new Object[] {LockType.Lease});
+
+    return args;
+  }
+
+  private final ApiClient apiClient;
+  private final LockType lockType;
+
+  public LeaderElectorTest(LockType lockType) {
+    try {
+      apiClient = ClientBuilder.defaultClient();
+    } catch (IOException ex) {
+      throw new RuntimeException("Couldn't create ApiClient", ex);
+    }
+    this.lockType = lockType;
+
+    // Lease resource requires special care with DateTime
+    if (lockType == LockType.Lease) {
```

Comment:
dang, the original kubernetes type requires a RFC3339 micro-sec timestamp here however joda-time wont support such precision anymore https://github.com/JodaOrg/joda-time/issues/139. i think we should consider getting https://github.com/kubernetes-client/java/pull/1418 merged so that serialization for micro-sec timestamp can work.

@himanshug can you add a TODO comment here saying:

```
TODO: switch date-time library so that micro-sec timestamp can be serialized in RFC3339 format w/ correct precision
```

Reply:
done... yeah `Lease` resource creation etc should just work out of the box :)

Reply:
that said, unrelated to this PR but, "serialization" code by default can either emit to ISO8601 or RFC3339 ... if k8s can accept RFC3339 for all the date-time fields in all the resources then we can switch the default serialization format in `JSON.java` to be RFC3339 or else `Lease` object serialization would need to be handled as a special case even after the switch somewhere.